gax-java/gax-httpjson/src/main/java/com/google/api/gax/httpjson/HttpJsonUnaryRequestParamCallable.java

```diff
@@ -0,0 +1,74 @@
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.httpjson;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.gax.rpc.ApiCallContext;
+import com.google.api.gax.rpc.RequestParamsExtractor;
+import com.google.api.gax.rpc.RequestUrlParamsEncoder;
+import com.google.api.gax.rpc.UnaryCallable;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * A {@code UnaryCallable} that extracts values from the fields of the request and inserts them into
+ * headers.
+ *
+ * <p>Package-private for internal usage.
+ */
+class HttpJsonUnaryRequestParamCallable<RequestT, ResponseT>
+    extends UnaryCallable<RequestT, ResponseT> {
+  private static final String DYNAMIC_ROUTING_HEADER_KEY = "x-goog-request-params";
+  private final UnaryCallable<RequestT, ResponseT> callable;
+  private final RequestUrlParamsEncoder<RequestT> paramsEncoder;
+
+  public HttpJsonUnaryRequestParamCallable(
+      UnaryCallable<RequestT, ResponseT> callable,
+      RequestParamsExtractor<RequestT> paramsExtractor) {
+    this.callable = Preconditions.checkNotNull(callable);
+    this.paramsEncoder =
+        new RequestUrlParamsEncoder<>(Preconditions.checkNotNull(paramsExtractor), false);
+  }
+
+  @Override
+  public ApiFuture<ResponseT> futureCall(RequestT request, ApiCallContext context) {
+    ApiCallContext newCallContext = context;
+    String encodedHeader = paramsEncoder.encode(request);
+    if (encodedHeader != null && !encodedHeader.isEmpty()) {
```

Comment:
Can you confirm with Slava? You are probably right and I agree that we should not send header if it's empty, but want to make sure it would not break existing behavior(in case some server are expecting empty header).

Reply:
Yep will do.

Reply:
Conversation with Slava:
```
the behaviour should be to NOT send the x-goog-request-params at all
server-side will likely do the right thing even if you send an empty value
```